src/porepy/numerics/nonlinear/convergence_check.py

```diff
@@ -0,0 +1,455 @@
+
+
+### Standard tolerances for assessing convergence.
+
+
+@dataclass
+class ConvergenceTolerance:
+    """Collection of standard tolerance for assessing convergence and divergence."""
+
+    tol_increment: float = np.inf
+    """Tolerance for increments for convergence."""
+    tol_residual: float = np.inf
+    """Tolerance for residuals for convergence."""
+    max_increment: float = np.inf
+    """Tolerance for increments for divergence."""
+    max_residual: float = np.inf
+    """Tolerance for residuals for divergence."""
+    max_iterations: int = 10000  # TODO replace with max int
+    """Upper bound for number of iterations for divergence."""
+
+
+### Base convergence criterion classes.
+
+
+class ConvergenceCriterion:
+    """Base class for convergence criteria.
+
+    Requires the implementation of the `_check` method to define the convergence.
+
+    """
+
+    def check(
+        self,
+        nonlinear_increment: float | dict[str, float],
+        residual: float | dict[str, float],
+        tol: ConvergenceTolerance,
+    ) -> tuple[ConvergenceStatus, ConvergenceInfo]:
+        """Check convergence.
+
+        Parameters:
+            nonlinear_increment: The increment in the solution variables from the
+                previous nonlinear iteration.
+            residual: The current residual vector of the nonlinear system.
+            tol: Collection of tolerances for assessing convergence and divergence.
+
+        Returns:
+            ConvergenceStatus: Convergence status of the non-linear iteration.
+            ConvergenceInfo: Information about the convergence check.
+
+        """
+        # Convert values to dict format
+        nonlinear_increment = self._make_dict(nonlinear_increment)
+        residual = self._make_dict(residual)
+
+        # Call the actual comparison
+        return self._check(nonlinear_increment, residual, tol)
+
+    def _check_dicts(self, dict1: dict[str, float], dict2: dict[str, float]) -> None:
+        """Check if two dictionaries have the same keys.
+
+        Parameters:
+            dict1: First dictionary to compare.
+            dict2: Second dictionary to compare.
+
+        Raises:
+            ValueError: If the dictionaries do not have the same keys.
+
+        """
+        if set(dict1.keys()) != set(dict2.keys()):
+            raise ValueError(
+                "Dictionaries do not have the same keys: "
+                f"{dict1.keys()} vs {dict2.keys()}"
+            )
+
+    def _make_dict(self, value: float | dict[str, float]) -> dict[str, float]:
+        """Convert a float or a dict to a unified dict format.
+
+        Parameters:
+            value: A float or a dict with string keys and float values.
+
+        Returns:
+            Float converted to a dict, or the original dict.
+
+        """
+        if isinstance(value, dict):
+            return value
+        else:
+            return {"": value}
+
+    @abstractmethod
```

Comment:
Please explain the relationship between the public and private check methods.

Reply:
There was a superfluous level. (Maybe there still is). Depending on how to deal with other criteria than "standard" relative or absolute criteria, see comment on `NanConvergenceCriterion`, the base class may be not required anymore. TBC after in-person discussion.